src/test/run-pass/union/union-nonzero.rs

```diff
@@ -0,0 +1,50 @@
+// run-pass
+#![allow(dead_code)]
+
+// Tests that unions aren't subject to unsafe non-zero/niche-filling optimizations.
+//
+// For example, if a union `U` can contain both a `&T` and a `*const T`, there's definitely no
+// bit-value that an `Option<U>` could reuse as `None`; this test makes sure that isn't done.
+//
+// Secondly, this tests the status quo to not apply such optimizations to types containing unions
+// even if they're theoretically possible. (discussion: https://github.com/rust-lang/rust/issues/36394)
+//
+// Notably this nails down part of the behavior that `MaybeUninit` assumes: that a
+// `Option<MaybeUninit<&u8>>` does not take advantage of non-zero optimization, and thus is a safe
```

Comment:
(Do we already?)

Reply:
I added a doctest for this case in https://github.com/rust-lang/rust/pull/60445/commits/99ac4b649042670f7294c551ca79c5355dc9ff29

Might be worthwhile adding explicit tests too, especially given that I could imagine `MaybeUninit` becoming a compiler built-in like `ManuallyDrop` in the future to support unsized types. But IMO that should happen separate to testing unions in general, precisely because `MaybeUninit` might not be a union one day.

Reply:
OK, good enough for now